pkg/controller/flash/flash.go

```diff
@@ -12,197 +12,85 @@
-	f.Add(errorKey, msg, vars...)
-}
-
-// ErrorNow adds a new error to the current flash instance.
-func (f *Flash) ErrorNow(msg string, vars ...interface{}) {
-	f.AddNow(errorKey, msg, vars...)
+	f.add(flashKeyError, msg, vars...)
 }
 
 // Errors returns the list of errors in flash, if any.
 func (f *Flash) Errors() []string {
-	return f.GetNow(errorKey)
+	return f.get(flashKeyError)
 }
 
 // Alert adds a new alert to the upcoming flash instance.
 func (f *Flash) Alert(msg string, vars ...interface{}) {
-	f.Add(alertKey, msg, vars...)
-}
-
-// AlertNow adds a new alert to the current flash instance.
-func (f *Flash) AlertNow(msg string, vars ...interface{}) {
-	f.AddNow(alertKey, msg, vars...)
+	f.add(flashKeyAlert, msg, vars...)
 }
 
 // Alerts returns the list of errors in flash, if any.
 func (f *Flash) Alerts() []string {
-	return f.GetNow(alertKey)
+	return f.get(flashKeyAlert)
 }
 
-// Add inserts the message into the upcoming flash for the given key.
-func (f *Flash) Add(key, msg string, vars ...interface{}) {
-	f.lock.Lock()
-	defer f.lock.Unlock()
-	f.next[key] = append(f.next[key], fmt.Sprintf(msg, vars...))
-
-	// Set the cookie on write.
-	http.SetCookie(f.w, &http.Cookie{
-		Name:   flashKey,
-		Value:  f.toCookieValue(f.next),
-		MaxAge: 300,
-		Path:   "/",
-	})
-}
-
-// AddNow inserts the message into the current flash for the given key.
-func (f *Flash) AddNow(key, msg string, vars ...interface{}) {
-	f.lock.Lock()
-	defer f.lock.Unlock()
-	f.now[key] = append(f.now[key], fmt.Sprintf(msg, vars...))
-}
-
-// Get returns the upcoming list of messages at the given key.
-func (f *Flash) Get(key string) []string {
-	f.lock.Lock()
-	defer f.lock.Unlock()
-	return f.next[key]
+// Clear removes all items from the flash. It's rare to call Clear since flashes
+// are cleared automatically upon reading.
+func (f *Flash) Clear() {
+	delete(f.values, flashKeyError)
+	delete(f.values, flashKeyAlert)
 }
 
-// GetNow returns the current list of messages at the given key.
-func (f *Flash) GetNow(key string) []string {
-	f.lock.Lock()
-	defer f.lock.Unlock()
-	return f.now[key]
+// add inserts the message into the upcoming flash for the given key.
+func (f *Flash) add(key flashKey, msg string, vars ...interface{}) {
+	var data []string
+	if v, ok := f.values[key]; ok {
+		data, _ = v.([]string)
+	}
+	f.values[key] = append(data, fmt.Sprintf(msg, vars...))
 }
 
-// toCookieValue converts the given map to its cookie value.
-func (f *Flash) toCookieValue(m map[string][]string) string {
-	b, err := json.Marshal(m)
-	if err != nil {
-		// This should never happen because JSON only fails to marshal when the
-		// values are unmarshallable (func, chan), or with circular references,
-		// neither of which can happen here.
-		log.Printf("failed to marshal as json: %v", err)
+// get returns the messages in the key, clearing the values stored at the key.
+func (f *Flash) get(key flashKey) []string {
+	if v, ok := f.values[key]; ok {
+		delete(f.values, key)
```

Comment:
Main difference in implementation: new flashes persist **until read** instead of **until the next request**.